### PR TITLE
Filter out ci_yaml auto roller targets from backfiller

### PR DESCRIPTION
**Pull request by ricardoamador**
Filter out the ci_yaml rollers from the backfiller.

*List which issues are fixed by this PR. You must list at least one issue.*
Fixes https://github.com/flutter/flutter/issues/127063

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read the [Flutter Style Guide] _recently_, and have followed its advice.
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart [hunk 1]
```diff
@@ -50,14 +50,28 @@ class BatchBackfiller extends RequestHandler {
     return Body.empty;
   }
 
+  // Remove the ci_yaml rollers from backfill list.
+  List<FullTask> removeCiYamlRollerFromBackFill(List<FullTask> taskList) {
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:54]: Can we control from .ci.yaml to make it easy to enable/disable? Say adding a property `backfill: false` to skip ones that we do not want to backfill.

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:54]: Yeah I guess so. Let me look into that. Do I need to then update a proto somewhere?

Comment (keyonghan) [reply app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:54]: No. Properties is already supported as in proto, and here we just add a sub element to the properties.

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:54]: Ah okay. Yeah I will do that then.

Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:54]: Okay that is a much better solution. Implemented this change.

File: app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart [hunk 2]
```diff
@@ -71,12 +71,16 @@ class BatchBackfiller extends RequestHandler {
       final FullTask task = taskColumn.first;
       final CiYaml ciYaml = await scheduler.getCiYaml(task.commit);
       final List<Target> ciYamlTargets = ciYaml.postsubmitTargets;
+
+      final List<Target> filteredCiYamlTargets = removeNonBackFillTargets(ciYamlTargets);
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:75]: can we move the logic to `ciYaml`, and directly obtain the list with `ciYaml.backfillTargets`?

File: app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart [hunk 3]
```diff
@@ -111,6 +115,26 @@ class BatchBackfiller extends RequestHandler {
     }
   }
 
+  // Remove targets that specify a property of 'backfill': 'false'. Otherwise we
+  // will assume that we include those targets when backfilling.
+  List<Target> removeNonBackFillTargets(List<Target> targets) {
+    final List<Target> filteredTargets = <Target>[];
+    for (Target target in targets) {
+      final Map<String, String> properties = target.value.properties;
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:123]: Instead we should use `target.getProperties()`.

File: app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart [hunk 4]
```diff
@@ -111,6 +115,26 @@ class BatchBackfiller extends RequestHandler {
...
+      if (properties.containsKey('backfill')) {
+        final bool? doBackFill = bool.tryParse(properties['backfill']!, caseSensitive: false);
+        // if properties does not contain backfill we assume backfill = true
+        if (doBackFill == null || doBackFill) {
+          filteredTargets.add(target);
+        }
+      } else {
+        // if no properties we assume backfill.
+        filteredTargets.add(target);
+      }
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:133]:
```suggestion
     if (!properties.containsKey('backfill') || (properties['release_build'] as bool)) {
       filteredTargets.add(target);
     }
```

File: app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart (hunk 2 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:75]: Sure. I can do that.

File: app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart (hunk 3 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart:123]: Sounds good.

File: app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart (hunk 4 above)
Comment (ricardoamador) [reply app_dart/lib/src/request_handlers/scheduler/batch_backfiller.dart]: Done.

File: app_dart/lib/src/model/ci_yaml/ci_yaml.dart [hunk 5]
```diff
@@ -83,6 +83,19 @@ class CiYaml {
     return enabledTargets;
   }
 
+  /// Filters post submit targets to remove targets we do not want backfilled.
+  List<Target> get backfillTargets {
+    final List<Target> postSubmitTargets = postsubmitTargets;
+    final List<Target> filteredTargets = <Target>[];
+    for (Target target in postSubmitTargets) {
```
Comment (keyonghan) [inline app_dart/lib/src/model/ci_yaml/ci_yaml.dart:90]:
```suggestion
    final List<Target> filteredTargets = <Target>[];
    for (Target target in postsubmitTargets) {
```

**Review by keyonghan — APPROVED** [review]
LGTM

**Pull request merged**